File: client-sdks/rust/sdk/command/impls/get_secondary_idx_two_command.rs
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// ------------------------------------------------------------------------------

use crate::comm::client_connector::ClientConnector;
use crate::command::command::Command;
use crate::command::command_type::CommandType;
use crate::model::vertex_cache_sdk_exception::VertexCacheSdkException;

pub struct GetSecondaryIdxTwoCommand {
    key: String,
    value: Option<String>,
    pub success: bool,
    pub status_message: String,
}

impl GetSecondaryIdxTwoCommand {
    pub fn new(key: String) -> Result<Self, VertexCacheSdkException> {
        if key.trim().is_empty() {
            return Err(VertexCacheSdkException::new(
                "GET By Secondary Index (idx2) command requires a non-empty key",
            ));
        }

        Ok(Self {
            key,
            value: None,
            success: false,
            status_message: String::new(),
        })
    }

    pub fn get_value(&self) -> Option<&str> {
        self.value.as_deref()
    }
}

impl Command for GetSecondaryIdxTwoCommand {
    fn execute(&mut self, connector: &mut ClientConnector) {
        let command_str = format!("GETIDX2 {}", self.key);

        match connector.send(&command_str) {
            Ok(response) => {
                let trimmed = response.trim();
                if trimmed.eq_ignore_ascii_case("(nil)") {
                    self.success = true;
                    self.status_message = "No matching key found, +(nil)".to_string();
                    self.value = None;
                } else if trimmed.starts_with("ERR") {
                    self.success = false;
                    self.status_message = format!("GETIDX2 failed: {}", trimmed);
                    self.value = None;
                } else {
                    self.success = true;
                    self.status_message = "OK".to_string();
                    self.value = Some(trimmed.to_string());
                }
            }
            Err(e) => {
                self.success = false;
                self.status_message = format!("GETIDX2 failed: {}", e);
                self.value = None;
            }
        }
    }

    fn command_type(&self) -> CommandType {
        CommandType::GetTertiaryIndex
    }

    fn is_success(&self) -> bool {
        self.success
    }

    fn status_message(&self) -> &str {
        &self.status_message
    }
}

File: client-sdks/go/sdk/command/get_secondary_idx_two_command.go
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// ------------------------------------------------------------------------------

package command

import (
	"strings"

	"github.com/vertexcache/vertexcache/client-sdks/go/sdk/comm"
	"github.com/vertexcache/vertexcache/client-sdks/go/sdk/model"
)

// GetSecondaryIdxTwoCommand handles the GETIDX2 command in VertexCache.
//
// Retrieves the value for a given key via the tertiary index (idx2).
// Returns an error if the key is missing or expired.
//
// Requires the client to have READ, READ_WRITE, or ADMIN access.
// This command supports idx2-based lookups only.
type GetSecondaryIdxTwoCommand struct {
	CommandBase
	key   string
	value string
}

// NewGetSecondaryIdxTwoCommand creates a new GetSecondaryIdxTwoCommand with the given key.
// Returns an error if the key is empty or invalid.
func NewGetSecondaryIdxTwoCommand(key string) (*GetSecondaryIdxTwoCommand, error) {
	if strings.TrimSpace(key) == "" {
		return nil, model.NewVertexCacheSdkException("GET By Secondary Index (idx2) command requires a non-empty key")
	}
	return &GetSecondaryIdxTwoCommand{key: key}, nil
}

// BuildCommand constructs the raw protocol string for the GETIDX2 command.
func (g *GetSecondaryIdxTwoCommand) BuildCommand() string {
	return "GETIDX2" + COMMAND_SPACER + g.key
}

// ParseResponse handles the raw response from the VertexCache server.
func (g *GetSecondaryIdxTwoCommand) ParseResponse(responseBody string) {
	if strings.EqualFold(responseBody, "(nil)") {
		g.SetSuccessWithResponse("No matching key found, +(nil)")
		return
	}

	if strings.HasPrefix(responseBody, "ERR") {
		g.SetFailure("GETIDX2 failed: " + responseBody)
	} else {
		g.value = responseBody
	}
}

// Value returns the value retrieved via idx2 lookup.
func (g *GetSecondaryIdxTwoCommand) Value() string {
	return g.value
}

// Execute runs the command using the shared ExecuteCommand helper.
func (g *GetSecondaryIdxTwoCommand) Execute(client *comm.ClientConnector) CommandInterface {
	return ExecuteCommand(g, client)
}
